src/core/qgsdatasourceuri.cpp

```diff
@@ -543,7 +543,7 @@ QString QgsDataSourceURI::uri( bool expandAuthConfig ) const
   if ( mWkbType != QGis::WKBUnknown && mWkbType != QGis::WKBNoGeometry )
   {
     theUri += " type=";
-    theUri += QgsWKBTypes::displayString( QgsWKBTypes::flatType(( QgsWKBTypes::Type( mWkbType ) ) ) );
+    theUri += QgsWKBTypes::displayString( QgsWKBTypes::Type( mWkbType ) );
```

Comment:
This worries me a little - looks like the decoding of type relies on QGis::fromNewWkbType, which has no handling of m dimensional types and returns an undefined value. I'm not sure what effect this could have.